compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs

```diff
@@ -416,12 +416,34 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                         _,
                     ) = pat.kind
                 {
-                    err.span_suggestion(
-                        upvar_ident.span,
-                        "consider changing this to be mutable",
-                        format!("mut {}", upvar_ident.name),
-                        Applicability::MachineApplicable,
-                    );
+                    if upvar_ident.name == kw::SelfLower {
+                        let mut fn_decl = None;
+                        let mut id_iter = self.infcx.tcx.hir().parent_iter(upvar_hir_id);
+                        while let Some((id, _)) = id_iter.next() {
+                            if let Some(fd) = self.infcx.tcx.hir().fn_decl_by_hir_id(id) {
+                                fn_decl = Some(fd);
+                                break;
+                            }
+                        }
+
+                        if let Some(fn_decl) = fn_decl {
+                            if !matches!(fn_decl.implicit_self, hir::ImplicitSelfKind::ImmRef | hir::ImplicitSelfKind::MutRef) {
+                                err.span_suggestion(
+                                    upvar_ident.span,
+                                    "consider changing this to be mutable",
+                                    format!("mut {}", upvar_ident.name),
+                                    Applicability::MachineApplicable,
+                                );
+                            }
+                        }
```

Comment:
These two loops can be collaprsed into one, I think.

```suggestion
                        for (_, node) in self.infcx.tcx.hir().parent_iter(upvar_hir_id) {
                            if let Some(fn_decl) = node.fn_decl() {
                                if !matches!(fn_decl.implicit_self, hir::ImplicitSelfKind::ImmRef | hir::ImplicitSelfKind::MutRef) {
                                    err.span_suggestion(
                                        upvar_ident.span,
                                        "consider changing this to be mutable",
                                        format!("mut {}", upvar_ident.name),
                                        Applicability::MachineApplicable,
                                    );
                                    break;
                                }
                            }
                        }
```